util/topgen/resets.py

```diff
@@ -0,0 +1,174 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Dict
+from .clocks import Clocks
+
+
+class ResetItem:
+    '''Individual resets'''
+    def __init__(self, raw: Dict[str, object], clocks: Clocks):
+        if not raw['name']:
+            raise ValueError('Reset has no name')
+
+        self.name = raw['name']
+        self.gen = raw.get('gen', True)
+        self.rst_type = raw.get('type', 'top')
+
+        # to be constructed later
+        self.domains = []
+        self.shadowed = False
+
+        self.parent = raw.get('parent', "")
+
+        # This can be a source clock or a derived source
+        if self.rst_type != 'ext':
+            self.clock = clocks.get_clock_by_name(raw['clk'])
+        else:
+            self.clock = None
+
+        self.sw = bool(raw.get('sw', 0))
+
+    def _asdict(self) -> Dict[str, object]:
+        ret = {
+            'name': self.name,
+            'gen': self.gen,
+            'type': self.rst_type,
+            'domains': self.domains,
+            'shadowed': self.shadowed,
+            'parent': self.parent,
+            'clock': self.clock,
+            'sw': self.sw
+        }
+
+        return ret
+
+
+class Resets:
+    '''Resets for the chip'''
+    def __init__(self, raw: Dict[str, object], clocks: Clocks):
+        self.hier_paths = {}
+        assert isinstance(raw['hier_paths'], dict)
+        for rst_src, path in raw['hier_paths'].items():
+            self.hier_paths[str(rst_src)] = str(path)
+
+        assert isinstance(raw['nodes'], list)
+
+        self.nodes = {}
+        for node in raw['nodes']:
+            assert isinstance(node, dict)
+            reset = ResetItem(node, clocks)
+            self.nodes[reset.name] = reset
+
+    def _asdict(self) -> Dict[str, object]:
+        ret = {
+            'hier_paths': self.hier_paths,
+        }
+
+        for name, node in self.nodes.items():
+            ret[name] = node._asdict()
+
+        return ret
+
+    def _get_reset_by_name(self, name: str) -> ResetItem:
+
+        ret = self.nodes.get(name, None)
+        if ret:
+            return ret
+        else:
+            raise ValueError('{name} is not a defined reset')
+
+    def set_reset_shadowed(self, name: str):
+        '''Mark particular reset as requiring shadow'''
```

Comment:
I'd consider calling this something like `mark_reset_shadowed`. The current name sounds like we might be setting a "reset_shadowed" flag, rather than setting a flag that says that a particular reset is shadowed.

Reply:
yeah i've removed this completely and am just doing a live check now.

Reply:
o wait i was thinking about something else.
But yes for this i agree `mark` is better. Will change.